### ISSUE TITLE
[Question] Scoped test suite contexts

### ISSUE THREAD

**Issue by idmadj** [microsoft/playwright#16789]
I need to test a pretty typical password-reset page. The page is only available after going through a relatively long flow: Go to sign-in page, click password reset link, enter email, click submit, open email, click link in email, arrive on tested page.

Multiple elements must be tested on this page: do a visual regression test, make sure the right error helpers appear, etc. Each element needs its own `test()`.

Multiple "suites" of tests need to be executed, some of which end by actually performing the password reset, which invalidates the page's URL and makes it unavailable to further testing.

Here are the options considered so far: 
- Using a non-scoped fixture, the initial flow would be repeated for each individual test, which is not efficient and would blow up testing time.
- Using a worker-scoped fixture, the initial flow is only done once, but can't be redone if the reset page is invalidated. Any test that runs after a password reset is performed will fail.
- Using a single `test()` with nested `test.step()`s works, but makes it impossible to selectively run specific steps (for instance using the VSCode extension).
- Using `test.beforeAll()` to set up the flow requires a common `page` object to be provided by a worker-scoped fixture, so that it can be used by subsequent `test()`s. This solution doesn't guarantee that the provided page is isolated and in a usable state. As it is shared across all tests, some might use it to sign in, then the password reset page would be unavailable. Relying on this sort of global, shared object will undoubtedly lead to test flakiness.

Is there anything obvious I'm missing? Having the ability to generate private, scoped, disposable contexts that can be used across a defined set of tests would be very helpful.

**Comment by yury-s** [comment]
First of all you need to make sure that the tests that may run in parallel do not interfere with each other (e.g. by invalidating the page's URL which may still be in use in other tests). Normally this is achieved by using different test accounts in different test suites. As I understood the some of the tests depends on the previously run tests, in which case you can use [serial mode](https://playwright.dev/docs/test-retries#serial-mode) and even reuse a page instance between tests as described in [this doc](https://playwright.dev/docs/test-retries#reuse-single-page-between-tests).

```ts

// test1.spec.ts
import { test } from '@playwright/test';

test.describe.configure({ mode: 'serial' });

const account1 = ...;
test.beforeAll(async () => { /* ... */ });
test('first', async ({ page }) => { /* ... */ });
test('second', async ({ page }) => { /* ... */ });
test('third', async ({ page }) => { /* ... */ });

// test2.spec.ts
import { test } from '@playwright/test';

test.describe.configure({ mode: 'serial' });

const account2 = ...;
test.beforeAll(async () => { /* ... */ });
test('first', async ({ page }) => { /* ... */ });
test('second', async ({ page }) => { /* ... */ });
test('third', async ({ page }) => { /* ... */ });

```
The account can be turned into a fixture which would provide a unique account name based on [worker index](https://playwright.dev/docs/test-parallel#worker-index-and-parallel-index), this way the tests can be isolated between parallel workers.

Would this work for you?

**Comment by yury-s** [comment]
> Multiple "suites" of tests need to be executed, some of which end by actually performing the password reset, which invalidates the page's URL and makes it unavailable to further testing.

Another option is to move them in a separate suite where the page setup is done in `beforeEach` so that each test works with its own copy?

**Comment by idmadj** [comment]
Thanks for your quick and detailed reply.

Using `beforeAll` with a shared variable does make sense. It avoids unnecessary reruns of the setup phase, shares context across specific tests and does provide isolation between suites, so that works!

If I want to organize this for code reuse, I can do this:
```js
// setup.mjs
export const getPasswordResetPage = async (browser) => {
  const page = await browser.newPage();
  // ... pick a unique account, go through the email flow, get to the password reset page...
  return page;
};

// test.spec.mjs
import {test} from '@playwright/test';
import {getPasswordResetPage} from './setup.mjs';

test.describe.serial("Password reset", () => {
  let passwordResetPage;

  test.beforeAll(async ({browser}) => {
    passwordResetPage = await getPasswordResetPage(browser);
  });

  test("Input field looks OK", () => {
    expect(await passwordResetPage.locator(`[data-test-id=input-password]`).screenshot())
    .toMatchSnapshot(`input-password.png`);
  });
});
```
Now what appears to a be a big downside is that `test.beforeAll` cannot access the `page` fixture. I have overridden the fixture to do some prerequisite navigation that's common to all tests in that file (as detailed in https://playwright.dev/docs/test-fixtures#overriding-fixtures). How would you encapsulate this behavior so that it can be used in `beforeAll`?

For instance:
```js
import {test as base} from '@playwright/test';
import {getPasswordResetPage} from './setup.mjs';

const test = base.extend({
  page: async ({page}, use) => {
    await page.goto('/signin');
    await page.click('[data-test-id=button-switchToPasswordReset]');
    await use(page);
  }
});

test.describe.serial("Password reset", () => {
  let passwordResetPage;

  test.beforeAll(async ({}) => {
    // How to access the `page` fixture here, so that it can be passed to `getPasswordResetPage` as a 
    // starting point, instead of creating a new page out of `browser` and repeating the initial navigation logic?
    passwordResetPage = await getPasswordResetPage(page);
  });

  test("Input field looks OK", () => {
    expect(await passwordResetPage.locator(`[data-test-id=input-password]`).screenshot())
    .toMatchSnapshot(`input-password.png`);
  });
});
```
I could move the navigation code that's in the fixture's body into a separate function, then also call it in `beforeAll`, but that seems needlessly convoluted. Would love to have your thoughts on this.

Another option to solve this would be to declare a fixture twice, once as a test-scoped fixture, and the other as a worker-scoped fixture, then use the worker-scoped one for tests that are not meant to incur a state change to the fixture, and the test-scoped one for those that do. This would allow for sufficient isolation when needed, context sharing in most cases, and a reduction in setup/teardown re-runs. That said, it would still cause some unnecessary re-runs, and the worker-scoped fixture would still not be able to access test-scoped fixtures like `page`, so not exactly great...

I've seen a few other people have suggested fixtures to be usable with test groups (#15542, #7482) and noticed that this functionality was under consideration. For what it's worth I think it would be a pertinent addition considering that the usage of fixtures seem to be encouraged over `beforeAll/beforeEach`. 

Using fixtures everywhere sounds like better DX too. Fixtures are arguably more conducive to code reuse, and I'd rather have setups/teardowns all expressed in the same way throughout a project than have some of them expressed as fixtures, and others as exported functions to be used in `beforeAll/afterAll`.

For instance, this is much neater than the examples above and allows for more flexible fixture composition:
```js
// fixtures.mjs
export const signInPage = async ({page}, use) => {
  await page.goto('/signin');
  await page.click('[data-test-id=button-password-reset]');
  await use(page);
}

export const passwordResetPage = async ({signInPage}, use) => {
  // ... start at `signInPage`, pick a unique account, go through the email flow, get to the password reset `page`...
  await use(page);
}

// test.spec.mjs
import {test as base} from '@playwright/test';
import {signInPage, passwordResetPage} from './fixture.mjs';

const test = base.extend({
  signInPage,
  passwordResetPage
});

test.describe("Password reset", ({passwordResetPage}) => { // ⬅️ where the magic happens!
  test("Input field looks OK", () => {
    expect(await passwordResetPage.locator(`[data-test-id=input-password]`).screenshot())
    .toMatchSnapshot(`input-password.png`);
  });
  test("Input only accepts valid characters", () => {
    // ... another assertion that uses `passwordResetPage`...
  });
});
```
Of course, group-scoped fixtures would need to be able to use test-scoped fixtures.

Thanks again for your help!

**Comment by yury-s** [comment]
> Now what appears to a be a big downside is that `test.beforeAll` cannot access the `page` fixture.

Yes, before all runs once for all tests in a worker, while `page` fixture is created before each test and destroyed after. You can prepare the page in `beforeEach` instead:

```ts
test.beforeEach(async ({page}) => {
  await page.goto('/signin');
  await page.click('[data-test-id=button-password-reset]');
});
```

> I have overridden the fixture to do some prerequisite navigation that's common to all tests in that file (as detailed in https://playwright.dev/docs/test-fixtures#overriding-fixtures). How would you encapsulate this behavior so that it can be used in `beforeAll`?

You don't have to use built-in page fixture and can reference `passwordResetPage` variable directly.

> I could move the navigation code that's in the fixture's body into a separate function, then also call it in beforeAll, but that seems needlessly convoluted. Would love to have your thoughts on this.

I believe you can turn `passwordResetPage`/'signInPage' in a worker fixture, something along these lines:

```ts
// fixture.ts
import { Page, test as base } from '@playwright/test';

// Note that we pass worker fixture types as a second template parameter.
export const test = base.extend<{}, { account: string; signedInPage: Page; }>({
  account: [ async ({ }, use) => {
    const username = 'account ' + process.env.TEST_WORKER_INDEX;
    await use(username);
  }, { scope: 'worker' }],

  signedInPage: [async ({ account, browser }, use) => {
    const context = await browser.newContext();
    const page = await context.newPage();
    // signin to the account ...
    (page as any).account = account;
    // Use signed-in page in the test.
    await use(page);
    await context.close();
  }, {scope: 'worker'}]
});
export { expect } from '@playwright/test';

// test1.spec.ts
import { test, expect } from './fixtures';

test('test 1', async({ signedInPage, account }) => {
  console.log(account);
  console.log((signedInPage as any).account)
});

// test2.spec.ts
import { test, expect } from './fixtures';

test('test 2', async({ signedInPage, account }) => {
  console.log(account);
  console.log((signedInPage as any).account)
});
```
Bear in mind though that such fixture are torn down only after all tests in the worker.

**Comment by idmadj** [comment]
At this point I'm fairly confident that you pointed me towards the most optimal solution using the current capabilities. 

> You can prepare the page in beforeEach instead

This is not possible as the page needs to be prepared before the code in `beforeAll` (the logic in `beforeAll` depends on the page as prepared by the overriden fixture).

> You don't have to use built-in page fixture and can reference passwordResetPage variable directly.

I actually do want to use the logic from an overridden page fixture in `beforeAll`. From what I understand, I would need to repeat any code that's in that fixture override into the `beforeAll` body to achieve that goal, which I'd rather avoid.

> I believe you can turn passwordResetPage/'signInPage' in a worker fixture, something along these lines
> ...
> Bear in mind though that such fixture are torn down only after all tests in the worker.

That wouldn't work as each test would perform some navigation on the used `page`, meaning that subsequent tests wouldn't necessarily begin from the same state.

This is another reason why I think the addition of group-scoped fixtures would be of great help. Generally, using `beforeAll` along with fixtures makes it hard to encapsulate and reuse code across a project as they ultimately use different paradigms despite serving similar purposes.

Let's close this issue at this point as I think we've got to the bottom of the problem. I'd just like to add my voice to those who believe that group-scoped fixtures would bring significant benefit to the elaboration of more complex tests. Let me know if I should open a separate feature request issue.

Thanks a lot for your help and support.

**Comment by yury-s** [comment]
We don't have plans to implement suite-scoped fixtures but let's keep the issue open for some time to see if it gets enough upvotes.

**Comment by GrayedFox** [comment]
Building off of these closed tickets: #15542 (test suite scoped fixtures) and #15665 (restart worker behaviour): all of our spec files are run sequentially and designed that way. I know it's discouraged but there are just too many benefits to managing our test suites that way (especially considering the time saved when running tests on BrowserStack).

Due to how the worker scoped fixture is called - once per worker (as one would expect) - this leaves a bit of a gap for anyone wishing to:

a) still use fixtures for code sharing and and init/setup/teardown
b) scope fixtures to individual spec files, but not to individual specs or individual workers

For serial specs this is particularly painful as state is intentionally kept between tests and context and browser closing and cleanup is all manual. Also although the fixture is working as it is described, debugging the following scenario was tricky:

Local machine
- 4 workers running 4 spec files
- causes a 1to1 mapping of workers to spec files, ensuring worker init code is done before each spec

Remote machine
- 2 workers running 4 spec files
- causes tests 1 (worker A), 2 (worker B), and 4 (worker B) to pass
- test 3 (worker A, numbering here refers to order of execution) fails due to being run on a worker that has not failed or been restarted
- I would expect test 4 to fail too but it seems like the very last test to run as part of a suite is a special case

For now rather than pass in an unused fixture to guarantee that init code is called for every spec no matter what/how many workers it is run on, I have simply removed the fixture entirely and call a `init()` method outside of the tests, before the PW runner has a chance to do things, which works for us it just means we have to do that in every spec file 🤷🏽

**Comment by pavelfeldman** [comment]
#### Why was this issue closed?

Thank you for your involvement. This issue was closed due to limited engagement (upvotes/activity), lack of recent activity, and insufficient actionability. To maintain a manageable database, we prioritize issues based on these factors.

If you have additional information not present in this issue that you think will help prioritizing it, please open a new issue and reference this one. More support or clarity on its necessity may prompt a review. Your understanding and cooperation are appreciated.
